File: CallbackTechniques/KeyValueObserving/KeyValueObserving.m
//
//  KeyValueObserving.m
//  CallbackTechniques
//
//  Created by Alexey Naumov on 30/01/2017.
//  Copyright © 2017 Alexey Naumov. All rights reserved.
//

#import "KeyValueObserving.h"

@implementation ObjCKeyValueObservingCaller

- (void)changeValue
{
  self.value += 1.;
}

@end

#define ObservedKeyPath   @"value"

@implementation ObjCKeyValueObservingCallee

- (void)startObservingObject:(id)object
{
  [object addObserver:self forKeyPath:ObservedKeyPath options:NSKeyValueObservingOptionNew context:NULL];
}

- (void)stopObservingObject:(id)object
{
  [object removeObserver:self forKeyPath:ObservedKeyPath];
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context
{
  if ([keyPath isEqualToString:ObservedKeyPath]) {
    self.wasCalled = YES;
  }
}

@end

File: CallbackTechniques/KeyValueObserving/KeyValueObserving.swift
//
//  KeyValueObserving.swift
//  CallbackTechniques
//
//  Created by Alexey Naumov on 30/01/2017.
//  Copyright © 2017 Alexey Naumov. All rights reserved.
//

import Foundation

class SwiftKeyValueObservingCallee : NSObject {
  
  var wasCalled = false
  private var observation: NSKeyValueObservation?
  
  func startObserving(object: SwiftKeyValueObservingCaller) {
    observation = object.observe(\.value, options: .new) { [weak self] (object, change) in
        self?.wasCalled = true
    }
  }
  
  func stopObserving(object: SwiftKeyValueObservingCaller) {
    observation?.invalidate()
    observation = nil
  }
}

class SwiftKeyValueObservingCaller : NSObject {
  
  @objc dynamic var value: CGFloat = 0
  
  func changeValue() {
    value += 1
  }
}
